### ISSUE TITLE
Subject Conversion Test: Stage 4 (names)

### ISSUE THREAD

**Issue by arwenhutt** [ucsdlib/damsmanager#190]
### Data input
File: 
[stage_4_data_input_test.xlsx](https://github.com/ucsdlib/damsmanager/files/1485635/stage_4_data_input_test.xlsx) (updated role values)

Includes columns:
1. old_ARK
2. old_predicate
3. new_ARK
4. new_predicate
5. new_role (values: arks for role, blank, or "subject" which indicates the name should be related to the object as a subject)
6. DAMS_ARK_Part_order
7. updated_label 

(6 and 7 are just for passing to the output report to help us with QA)
 
### Matching criteria
IN CLR/OLR records, match:
`old_predicate` AND `old_ARK`

### Process
1. IF there is an exact match on all criteria (AND/&&)
	1. AND IF name is linked via a relationship blank node OR new_role != NULL
		1. AND IF new_role != NULL
			1. THEN check record for new_predicate AND new_ARK value AND new_role value
				1. IF match is found
					1. THEN do nothing
				2. ELSE
					1. Add new_ark using new_predicate and new_role 
		2. ELSE IF new_role = NULL
			1. Store role value from DAMS relationship node
			2. THEN check record for new_predicate AND new_ARK value AND stored role value
				1. IF match is found
					1. THEN do nothing
				2. ELSE
					1. Add new_ark using new_predicate and stored role
	2. ELSE (IF name is NOT linked via a relationship blank node, AND does not have a new_role value)
		1. THEN check record for new_predicate AND new_ARK value
			1. IF match is found
				1. THEN do nothing
			2. ELSE
				1. Add new_ark using new_predicate

### Report of results
1. object ark where match was made
2. new_ARK added
3. new_predicate added
4. new_role added
5. old_ARK
6. old_predicate
7. old role value (when relevant)
8. updated_label (from data input file)
9. DAMS_ARK_Part_order (from data input file)

### Related work
Subject Conversion Test: Stage 3 (linking non-names) https://github.com/ucsdlib/damsmanager/issues/184
Subject Conversion Test: Stage 2 https://github.com/ucsdlib/damsmanager/issues/178
Subject Conversion Test: Stage 1 https://github.com/ucsdlib/damsmanager/issues/177
Overall subject conversion process https://github.com/ucsdlib/damspas/issues/355

**Comment by GregReser** [comment]
I think the logic is right.
Add to the report: updated_label.  I find it helpful when checking the results

**Comment by remerjohnson** [comment]
The final `else` clause in "ii": I'm not clear on what conditional it is working off of, but as far as I can tell, it all might work

**Comment by lsitu** [comment]
@arwenhutt Is the ticket ready? I see a dummy link for the Data File: stage_4_data_input_test.xlsx. Could you attach it again? Thanks.
For the report, I think it'll be lengthy. We may get multiple occurrences in the report for each name that are matched in a CLR/OLR. Does it what you want?

**Comment by arwenhutt** [comment]
@lsitu I’m working on the file now - should have it ready pretty soon!

**Comment by arwenhutt** [comment]
Thanks all, 
@GregReser I added updated_label, as well as a few other things from the data input file/test that may help with QA

@remerjohnson I added a paranthetical to clarify the ELSE scenario (it's basically if the condition in 1. is met but not that in 1.i.)

@lsitu for the report, it's fine if it's lengthy and there's duplicates, I think it will be useful for us.  My only concern is if it will be very difficult for you to code.  

I've attached the file, and I think this is all ready to go!  Thanks!

**Comment by lsitu** [comment]
@arwenhutt It looks good. I am working on the logic for the tool now.

**Comment by lsitu** [comment]
@arwenhutt I see the values in new_role column are literal that may need translation. For example:
Row#474 need to fix order 
Row#475 need to fix order 
Row#476 role, former owner?
Row#4535 subject
Row#4779 dams:topic
Row#5272 dams:corporate:Vessel
...

IN 1.i, we may need to use the new_role value to find a match and/or create a new relationships. Could it be the ARK for the role? Or at least a label of a role that we can look up from DAMS repository?

**Comment by arwenhutt** [comment]
@lsitu oops! yup, that needs fixing.  I think we should be able to look up the arks since we have a report of them.  I'm not sure I'll have time to do it tonight, but will try and will let you know when it's done.

**Comment by lsitu** [comment]
Thanks @arwenhutt .

**Comment by arwenhutt** [comment]
@lsitu I've updated and reattached the input file.  
Role values now include: arks for role (just vessel and cruise), blank, or "subject" which indicates the name should be related to the object as a subject

**Comment by lsitu** [comment]
@arwenhutt Thanks. It looks like there are still some non-uri values like "subject" for new_role, should I simply ignore those non-uri values in the new_role column?  For example, row#4776 - 4779.

**Comment by arwenhutt** [comment]
@lsitu the ones with "subject" should be linked to the object directly (instead of via a relationship value), I'm happy to move this if there's a better way to express this.

**Comment by lsitu** [comment]
@arwenhutt I am running a test that simply ignores those subject/non-url values at this time. So I think it should be fine if those subject values are not roles.

**Comment by arwenhutt** [comment]
They aren't roles but they may indicate a change in how the name is related to the object, so we may need to modify the processing instructions and/or the data to account for them.

**Comment by lsitu** [comment]
@arwenhutt Thanks for clarifying that. I've got the report as a test that you can reference for modifying the processing instructions and/or the data. But it's 32M in size and I can't attach it here. Would you like to take a look?

Also, another report that is similar to the one in staging 3, which count for the changes with columns like (Records matched | Relationships added to | Subjects added to | Relationships exist | Subjects exist) in row base:
[report-matched-test.txt](https://github.com/ucsdlib/damsmanager/files/1488908/report-matched-test.txt)

There will be total 53660 CLRs/OLRs affected in this stage. Please let me know once the ticket is ready for me. Thanks.

**Comment by arwenhutt** [comment]
@lsitu   I thought for a second you meant it had 32 million rows (which it might!) heh.  
Yeah, I should look at it, just let me know where's convenient to put/get it. Thanks!

**Comment by lsitu** [comment]
@arwenhutt Aha, it's 32 megabytes with 124260 rows only. I notice some rows may be duplicated since the report is generated basing on the query result of old_ARK and old_predicate in each row. However, I can eliminate those duplicate rows in the final report.
I zipped it and the size is much smaller now:
[report-stage4-test.zip](https://github.com/ucsdlib/damsmanager/files/1489145/report-stage4-test.zip)

**Comment by arwenhutt** [comment]
@lsitu There are some things that I think need changing:

## 1 Roles are being changed that shouldn't
I'm seeing a lot of rows where it looks like the roles is being changed when it should not be.   

For example:
For DAMS_ARK_Part_order = http://library.ucsd.edu/ark:/20775/bb6588466h
according to the data input spreadsheet - there is no new role value indicated and so the old role should be retained:

old_ARK | new_ARK | old_predicate | new_predicate | new_role | updated_label | DAMS_ARK_Part_order
-- | -- | -- | -- | -- | -- | --
http://library.ucsd.edu/ark:/20775/bb6588466h | http://library.ucsd.edu/ark:/20775/bb06839775 | dams:corporateName | dams:corporateName |   | University of   California, San Diego | http://library.ucsd.edu/ark:/20775/bb6588466h

From the report-stage4-test.txt, the old role (creator) is changed to (distributor)

Object   ARK | new_ARK   added | new_predicate   added | new_role   added | old_ARK | old_predicate | old   role value | updated_label | DAMS_ARK_Part_order
-- | -- | -- | -- | -- | -- | -- | -- | --
http://library.ucsd.edu/ark:/20775/bb6110371m | http://library.ucsd.edu/ark:/20775/bb06839775 | dams:corporateName | http://library.ucsd.edu/ark:/20775/bb5121025d | http://library.ucsd.edu/ark:/20775/bb6588466h | dams:corporateName | http://library.ucsd.edu/ark:/20775/bb67931665 | University   of California, San Diego | http://library.ucsd.edu/ark:/20775/bb6588466h

It looks like about 5,000k rows in the full report show this pattern of the role being changed that shouldn't be.


## 2 new role values off in report-matched-test.txt 
In report-matched-test.txt, the column labeled "new_role added" seems to contain the new_ARK values, not role values

## 3 no change match processing and action reported
In both reports, every match seems to be recorded as triggering the addition of a new_predicate and (when relevant) a new role, even when the condition in 1.i.a.a. are met (everything is the same) and there should be no change.  We have four actions which may result from the processing:
* No change (1.i.a.a.a.a., 1.i.b.b.a.a, & 1.ii.a.a.a.)
* Addition using new_ark, new_predicate, and new_role (1.i.a.a.b.a)
* Addition using new_ark, new_predicate, and stored (old/original) role (1.i.b.b.b.a.)
* Addition using new_ark & new_predicate (1.ii.a.b.a.)

I am not sure whether this is a problem with the matching criteria, algorithm, or a function of the design of the report.  

I suggest adding an additional column to the output report indicating action taken:
* no change
* added with new role
* added with old role
* added with no role 

Related: we should probably modify the column headers for 2-4 for clarity to: new_ARK, new_predicate, and new_role, and they should contain the same values as the data input file

## 4  report indicate how name is linked
It would also be very useful to have an indication in the report of how the name is related to the object - I think this could be the predicate linked to the object, examples would be: dams:relationship, dams:personalName, dams:rightsHolderPersonal, etc.  
	
In the cases of names linked as subjects (i.e. dams:personalName, dams:corporateName, etc.) this would I think be the same as "new_predicate added", but it will still be important for indicating how the name is being used.  


With the changes proposed in 3 & 4 the updated report spec would then be:
1. object ark where match was made
2. action (no change, added with new role, added with old role, added with no role) 
3. linked by (i.e. dams:personalName, dams:relationship, dams:rightsHolderCorporate, etc.)
4. new_ARK 
5. new_predicate
6. new_role
7. old_ARK
8. old_predicate
9. old role value (when relevant)
10. updated_label (from data input file)
11. DAMS_ARK_Part_order (from data input file)

**Comment by lsitu** [comment]
@arwenhutt Attached is the updated report basing on your feedback above:
[report-stage4.zip](https://github.com/ucsdlib/damsmanager/files/1492770/report-stage4.zip)

And the one similar to stage 3:
[report-matched.txt](https://github.com/ucsdlib/damsmanager/files/1492780/report-matched.txt)

**Comment by lsitu** [comment]
@arwenhutt Do we need to modify the processing instructions and/or the data for those "subject" values in the new_role column? AT this time, I simply ignored those non-url values like "subject" to generate the reports above. And those rows with "no change" action value in the report are name subject links that are already added to the CLRs/OLRs (duplicates).

Please let me know if you think we are ready to update the CLRs/OLRs.

**Comment by arwenhutt** [comment]
Thanks @lsitu 
I haven't had a chance to look at the report yet, and have a few things I have to take care of first so it'll probably be a few hours before I have time to dig into it.  I'll let you know as soon as I can how it looks.

For your question about subjects - I'm not sure if this is being covered in the current code or if it needs to be modified.  I think it is probably okay if all of the rows (from the input spreadsheet) with a value of "subject" in new role, have in the associated (using the DAMS_ARK_part_order as key) have no value in "new_role added" and have in the "action" column, either "added with no role" or "no change".  Do you think this is true for the new_role=subject rows?

**Comment by lsitu** [comment]
@arwenhutt I have no ideas regarding whether the action of the rows with a value of "subject" in the new_role column are "added with no role" or "no change" or not. Do you need another report to include a column for it?

**Comment by arwenhutt** [comment]
@lsitu How about this:  How does the processing code determine how to link a name to the object in question?

**Comment by lsitu** [comment]
@arwenhutt First look up the objects that match `old_predicate` AND `old_ARK` in each row, then for each object found, find all links that match the`old_predicate` AND `old_ARK` and apply the rules in the `Process` section to determine whether a dams:Relationship or subject name need to be added to the object or not. The only difference is that those non-url values in the new_role column like "subject" will be treated as blank value (new_role == NULL).

**Comment by arwenhutt** [comment]
@lsitu okay, I'm still not sure just reading the description.  But I finished (well, enough) of my other thing and will start looking at the report to see if that will help me tell.  It may be though that some of this we won't know for sure until we go through the real test.

**Comment by lsitu** [comment]
@arwenhutt Thanks. Just let me know what's next once you are done with reviewing the report.

**Comment by arwenhutt** [comment]
@lsitu issues/questions below - I've continued the numbering from the previous comment so it won't be confusing when we need to refer to any of them:

## 5 
A number of rows with action="no change" should be triggering a change (e.g. the new_ark does not match the old_ark or the new_predicate does not match the old_predicate)


## 6 
I'm seeing rows that seem like they should not require any addition (the new and old arks, predicates, and roles all match) that are show action="added with old role"


## 7 
It looks like names linked by rightsHolderPersonal or Corporate (and perhaps other predicates I haven't caught) aren't being caught by the processing.  

For example from the data_input file:

old_ARK | new_ARK | old_predicate | new_predicate | new_role | updated_label | DAMS_ARK_Part_order
-- | -- | -- | -- | -- | -- | --
http://library.ucsd.edu/ark:/20775/bb2697785c | http://library.ucsd.edu/ark:/20775/bb2697785c | dams:corporateName | dams:corporateName |   | State of   California, Department of Fish and Game | http://library.ucsd.edu/ark:/20775/bb2697785c

According to the original data dump - this combo was present in 46 objects, and looking at the example link provided https://library.ucsd.edu/dc/collection/bb07861424/rdf it is still present in at least one record, encoded as:

<dams:rightsHolderCorporate>
<mads:CorporateName rdf:about="http://library.ucsd.edu/ark:/20775/bb2697785c">
<mads:authoritativeLabel>State of California, Department of Fish and Game</mads:authoritativeLabel>
<mads:elementList rdf:parseType="Collection">
<mads:NameElement>
<mads:elementValue>State of California, Department of Fish and Game</mads:elementValue>
</mads:NameElement>
</mads:elementList>
</mads:CorporateName>
</dams:rightsHolderCorporate>

But the report-matched.txt says that it did not match any records.
There are about 1000 rows that show no match, which is too high.  Almost everything in the data input file should match at least one record in the DAMS (since the data input file is derived from data output from the DAMS records).  There may be a small number that have been manually edited since the original reports but that should be a much smaller number.

(another indication that rights holders aren't being caught is that there are no rights holder values in the "linked by" field)

## 8
I'm unclear on what is included in the detailed report. Only about 30% of the rows from the input spreadsheet seem to be represented.  The report-matched.txt report seems to line up 1:1 with the data input file.  As with above, almost everything should match even if no change is triggered.

## 9 
A clarification on the "linked by" field in the report, can this be changed to show how the additions are linked? It would be fine to not provide a value for those which have "no change" since we can infer the link from other data.  This would help us a lot in QA.

## 10 
In the detailed report, for the columns "new_ARK added", "new_predicate added" and "new_role added" - the column header and the values provided seem to conflict.  All of these columns contain values even when there is no change needed. I am not sure if this is a data and column header mismatch, or related to other problems in the processing or mapping.  We need to have a better idea of what the values indicate and have a column header to represent that.

## 11 
These are just some formatting things but worth noting:
1. There are character encoding problems with the labels, it's not as essential but would still be helpful to have them output correctly
2. Can the arks be returned as functional urls that take us to the rdf view? This would be very helpful for us and improve the speed of our review.

**Comment by lsitu** [comment]
@arwenhutt: _Sorry I may need to clarify that only those objects that are matched and need action like `added with new role`, `added with old role`, `added with no role` will be included in the report. Those actions with `no change` are just dups that are already done previously:_

## 5 
A number of rows with action="no change" should be triggering a change (e.g. the new_ark does not match the old_ark or the new_predicate does not match the old_predicate)

_A: As clarify ing above, the report only include those objects that need some actions to add a dams:Relationship or a name subject. However, I found lots of links already added in the previous rows, so I just put action="no change" for those rows. Maybe terms like action="Already added with new role" is better in this case?_

## 6 
I'm seeing rows that seem like they should not require any addition (the new and old arks, predicates, and roles all match) that are show action="added with old role"

_A: There should be some gaps her since I believe those rows should match the conditions above for action="added with old role". But Could you show me an example so that we can examine it to see what's the issue?_

## 7 
It looks like names linked by rightsHolderPersonal or Corporate (and perhaps other predicates I haven't caught) aren't being caught by the processing.  

_A: As the example showing, the old_predicate is `dams:rightsHolderCorporate` but I don't see any values like dams:rightsHolderCorporate in the old_predicate column. I only see some values like `dams:rightsHolderName`. Maybe we need to correct it in the source input file stage_4_data_input_test.xlsx?_

For example from the data_input file:

old_ARK | new_ARK | old_predicate | new_predicate | new_role | updated_label | DAMS_ARK_Part_order
-- | -- | -- | -- | -- | -- | --
http://library.ucsd.edu/ark:/20775/bb2697785c | http://library.ucsd.edu/ark:/20775/bb2697785c | dams:corporateName | dams:corporateName |   | State of   California, Department of Fish and Game | http://library.ucsd.edu/ark:/20775/bb2697785c

According to the original data dump - this combo was present in 46 objects, and looking at the example link provided https://library.ucsd.edu/dc/collection/bb07861424/rdf it is still present in at least one record, encoded as:

<dams:rightsHolderCorporate>
<mads:CorporateName rdf:about="http://library.ucsd.edu/ark:/20775/bb2697785c">
<mads:authoritativeLabel>State of California, Department of Fish and Game</mads:authoritativeLabel>
<mads:elementList rdf:parseType="Collection">
<mads:NameElement>
<mads:elementValue>State of California, Department of Fish and Game</mads:elementValue>
</mads:NameElement>
</mads:elementList>
</mads:CorporateName>
</dams:rightsHolderCorporate>

But the report-matched.txt says that it did not match any records.
There are about 1000 rows that show no match, which is too high.  Almost everything in the data input file should match at least one record in the DAMS (since the data input file is derived from data output from the DAMS records).  There may be a small number that have been manually edited since the original reports but that should be a much smaller number.

(another indication that rights holders aren't being caught is that there are no rights holder values in the "linked by" field)

## 8
I'm unclear on what is included in the detailed report. Only about 30% of the rows from the input spreadsheet seem to be represented.  The report-matched.txt report seems to line up 1:1 with the data input file.  As with above, almost everything should match even if no change is triggered.
_A: I only include those objects that need `action` in the report as I clarifying above. But it seems like you expect to see all records that are matched. Correct?_

## 9 
A clarification on the "linked by" field in the report, can this be changed to show how the additions are linked? It would be fine to not provide a value for those which have "no change" since we can infer the link from other data.  This would help us a lot in QA.

_A: Yes, I think we can do it next time when we include those links that are already exist in a CLR/OLR._

## 10 
In the detailed report, for the columns "new_ARK added", "new_predicate added" and "new_role added" - the column header and the values provided seem to conflict.  All of these columns contain values even when there is no change needed. I am not sure if this is a data and column header mismatch, or related to other problems in the processing or mapping.  We need to have a better idea of what the values indicate and have a column header to represent that.

_A: Basing on the rules above, this row need to take a action to `add with new role`, `add with old role`, or `add with no role`. However, I found these links are added earlier in other rows. So I just mark it as `no action` for now. Maybe we need to use other terms as the action in this case?_

## 11 
These are just some formatting things but worth noting:
1. There are character encoding problems with the labels, it's not as essential but would still be helpful to have them output correctly

_A: I see it correct when I open it with a text editor like Notepad++. Could you import it to Excel with UTF-8 encoding? Please let me know if you still see it corrupted and we can find a way to deal with it. Thanks._

2. Can the arks be returned as functional urls that take us to the rdf view? This would be very helpful for us and improve the speed of our review.
_A: Sure. I think we can use the dams API URL next time._

**Comment by lsitu** [comment]
@arwenhutt Basing on your feedback and my response above,  I've moved ahead to generate the report with the following changes in the  process and the reports:
1. To catch the rights holders names, an additional row with old_predicate=dams:rightsHolderxxx is added to the data file stage_4_data_input_test.xlsx for every name row with old_predicate= () :
dams:personalName => dams:rightsHolderPersonal
dams:corporateName => dams:rightsHolderCorporate
 dams:familyName => dams:rightsHolderFamily
dams:conferenceName => dams:rightsHolderConference
dams:rightsHolderName => dams:rightsHolder (DAMS 4 has no predicates like dams:rightsHolderName but dams:rightsHolder)

Note: if there is no matches found for the rights holder row added, it will not added to report-matched.txt so that we only got those rows that have matched results.

2. Update `action` to include the following values:
no change - matched new predicate, new_ark, and new_role and no action is taken
added with new role
already added with new role - already added with new role in previous rows.
added with old role
already added with old role e - already added with old role in previous rows.
added with no role
already added with no role e - already added with no role in previous rows.

3. Report all results found instead of just those CLRs/OLRs that need `action` in the last report.

4. Use DAMS API url that will load the RDF/XML metadata for "object ARK'.

Here is  the report for stage 4:
[report-stage4.zip](https://github.com/ucsdlib/damsmanager/files/1506660/report-stage4.zip)

And the report for records matched:
[report-matched.txt](https://github.com/ucsdlib/damsmanager/files/1506657/report-matched.txt)

**Comment by arwenhutt** [comment]
@lsitu I haven't had as much time with this as I would like today, but I feel like things are getting less clear for me as things go on.  I think we may get further if we can separate the review of the processing code from review of the report creation.  

To evaluate how the processing is occurring I would like to:
1. look at the processing code - I might not be able to understand it all, but I may be able to follow the logic well enough to gain some confidence that we are on the same page or understanding of where there are differences
2. ask @mcritchlow if he has the bandwidth to evaluate the code against the requirements
3. run a test batch of selected records

Once we feel confident that the code modification is working as needed, then we can look at what kind of report we can generate.

**Comment by jessicahilt** [comment]
@lsitu It looks like you've done a great job getting this project most of the way there! Can you push the code up to github so we can review the code? Thanks!

**Comment by lsitu** [comment]
@arwenhutt I think we could create a private repo for it if you want to see the codes, or I can remove the credentials and attach the source file related to the logic here. Which way would you like?

Could you tell me what become less clear per your view? I believe the gaps between us are how to catch those links for rights holder names are caught,  and why there are some links are added previously (action: already added with ...).
So the changes above are just to add additional rows to the input data to catch those rights holder names, which is missing from the process above. And if any links with new_predicate, new_ARK and/or new_role (dams:Relationship) have already added in other rows previously, we just mark it as "already added with ...".

Running a test batch of selected records sounds good to me since it's a simple way to catch any issues that might have.

I think another way that we can do is to query all CLRS/OLRS that contain links with old_ARK (compare to match by old_predicate and old_ARK), then apply the logic to determine what action we need by the old_predicate and/or linked by predicate.

@mcritchlow What do you think?

**Comment by gamontoya** [comment]
@lsitu In a meeting with Arwen and Matt. Can you just run the code on QA right now? We are going to have multiple iterations of this work even if it takes weeks, but this will better help Arwen rather than having to review report.

**Comment by lsitu** [comment]
@gamontoya Sure. I am moving ahead to update the CLRs/OLRs now. Thank you all.

**Comment by lsitu** [comment]
@arwenhutt I think we've done with stage 4 to update the CLRs/OLRs. 
Attached is the report:
[report-stage4.zip](https://github.com/ucsdlib/damsmanager/files/1519623/report-stage4.zip)

And the one for matched counts:
[report-matched.txt](https://github.com/ucsdlib/damsmanager/files/1519626/report-matched.txt)

But the SOLR re-indexing process is still on-going with more than 46,000 left in the queue at this time, which could be done by early next Monday. I'll let you know once it finishes.

**Comment by lsitu** [comment]
@arwenhutt I just checked the solr re-indexing process. It's delayed and there are still more than 20,000 items in the queue http://lib-hydratail-qa.ucsd.edu:8161/admin/queues.jsp. It looks like it could take another day and a half or two.

**Comment by lsitu** [comment]
@rstanonik It looks like the Solrizer on lib-hydratail-qa is stopped: http://lib-hydratail-qa.ucsd.edu:8161/admin/queues.jsp. Could you take a look? Thanks.

**Comment by lsitu** [comment]
@rstanonik Have you got a chance to look at DAMS Solrizer on QA? It'll be nice if we can restart it for now so that it won't delay the SOLR indexing process. Thanks.

**Comment by rstanonik** [comment]
I restarted it.
I'll create a jenkins plan, so you can restart it too, if need be.

**Comment by lsitu** [comment]
@rstanonik That sounds great! Thanks a lot.

**Comment by lsitu** [comment]
@arwenhutt The SOLR re-indexing process is finished on QA. I think it's ready for you to review the metadata changes in Stage 4 now. Thanks.

**Comment by arwenhutt** [comment]
@lsitu thanks!  I should have some time tomorrow to start looking at it.

**Comment by GregReser** [comment]
@lsitu Should we be seeing all the changes from this test in http://libraryqa.ucsd.edu/dc/...?

In http://libraryqa.ucsd.edu/dc/object/bb0138369g , I don't see the new topic bb00356037 "Correspondence".  I see it in the RDF.  Perhaps the indexing is not complete?
![image](https://user-images.githubusercontent.com/5453716/33789020-a946c94c-dc2a-11e7-9a7e-7a67a844a3d5.png)

![image](https://user-images.githubusercontent.com/5453716/33789038-cb8d995e-dc2a-11e7-900f-3835e57cec4e.png)

**Comment by lsitu** [comment]
@arwenhutt:   @rstanonik and I just found that there some issues with SOLR in QA and the Solrizer failed to index the records. I am moving ahead to reindex the records updated in this stage now. So you may only be able to review the metadata at this time but not through the object viewer in damspas. Sorry about it.

**Comment by arwenhutt** [comment]
Jinx!

**Comment by GregReser** [comment]
@lsitu funny, we fond this at the same time!

**Comment by lsitu** [comment]
@GregReser Yep. Ron and I are working together to see what's going wrong with SOLR for a little bit while.

**Comment by lsitu** [comment]
@arwenhutt @GregReser Do you have a list of objects that you want to review? I can index those objets for you to review first. This will save us some time to wait for SOLR index to be done.

**Comment by GregReser** [comment]
@lsitu yes, that makes sense.  Let's start with these:
bb0138369g
bb1536977g
bb46769317
bb21170551
bb4404344d
bb5291268s
bb56324513

**Comment by lsitu** [comment]
@GregReser All 7 records above are updated in SOLR now. So you could review them in the object view like http://libraryqa.ucsd.edu/dc/object/bb0138369g at this time. 

Would you like me to move ahead to index all other records now?

**Comment by arwenhutt** [comment]
@lsitu yes please

**Comment by lsitu** [comment]
@arwenhutt Okay. All CLRs/OLRs updated in stage 4 are added to the solr index queue now. I think it could be done by next Thursday and you also can track the status [here](http://lib-hydratail-qa.ucsd.edu:8161/admin/queues.jsp). The number under column "Number Of Pending Messages" is the number of objects still need to be done. 
I'll be out next sprint, but I'll update the status for you once I can.

**Comment by lsitu** [comment]
@arwenhutt SOLR re-indexing is done now. I think it's ready for you to review now. Thanks.

**Comment by lsitu** [comment]
@arwenhutt Anything else do I need to do for stage 4?

**Comment by arwenhutt** [comment]
@lsitu we are still reviewing - we've found a few things that we're tracing back to identify what's happening.  We'll let you know as soon as we have more specifics.

**Comment by arwenhutt** [comment]
@lsitu there are some problems with rights holders names being added as subjects.  I can rewrite the processing instructions to more explicitly include these, but I need a little more information so they are correct:

1. There are three ways I've found rights holders encoded (below).  Are there any other ways  rights holders are encoded and linked to records in the DAMS?
    1. dams:rightsHolderName/mads:name
    2. dams:rightsHolderCorporate/mads:CorporateName
    3. dams:rightsHolderPersonal/mads:PersonalName
2. A report of all uses of dams:name/mads:Name linked to a record (example of the encoding in this record https://library.ucsd.edu/dc/collection/bb1605660t/rdf) in the dams.  Report contents: "record ark", "name ark" and "display label"

 ( @jessicahilt let me know if I should create a new ticket for either of these : )

Thanks!

**Comment by lsitu** [comment]
@arwenhutt:
For # 1, theoretically, I think rights holders could be encoded with other predicates like dams:rightsHolder, dams:rightsHolderConference, and dams:rightsHolderFamily since I see these predicates are defined in dams4. But I can't find any records that are lined by these predicates. So I think we can ignore them for now.
For # 2, regarding the report for dams:name/mads:Name, I am running a process for it now and I'll let you know once it's done.

**Comment by lsitu** [comment]
@arwenhutt Here is the report for # 2, the uses of dams:name/mads:Name linked to records in dams. I added an additional column "linked by" to include the predicate and name class path. Please let me know if you need anything else.
[report_dams_name.txt](https://github.com/ucsdlib/damsmanager/files/1616965/report_dams_name.txt)

**Comment by arwenhutt** [comment]
@lsitu thanks, that helped.  The names are gnarly.  
Basically we need to add logic to the code to identify names linked as rights holders and add (when there is a change in ark) them with the new ark but retaining the rights holder relationship.  
But since the rights names have already been added as subjects to objects, I think we'll need to start another test round.  I believe @gamontoya is going to set up a ticket to resync QA data and I'll start working on new tickets/input files updated based on this round of testing on Monday.  I think I can close this ticket now.
